deprecate back argument in attacks

npapernot opened the pull request:
Closes #631 

(WIP)

goodfeli commented on cleverhans/attacks.py:
```diff
@@ -21,22 +21,30 @@ class Attack(object):
 
+    self.tf_dtype = tf.as_dtype(dtypestr)
     self.np_dtype = np.dtype(dtypestr)
 
```
Add a check on the type of `sess`. This will also make sure that an old positional argument passed for `back` doesn't end up going to `sess` now that `back` has been removed.

goodfeli commented:
OK to merge after adding the check on type of `sess`

goodfeli commented on cleverhans/attacks.py:
```diff
@@ -47,7 +57,7 @@ def __init__(self, model, back='tf', sess=None, dtypestr='float32'):
     # Prepare attributes
     self.model = model
-    self.back = back
+    self.back = 'tf'
```
Do we still need self.back at all?

goodfeli replied:
If we need self.back for backwards compatibility purposes, maybe we should make it a python property ( https://www.programiz.com/python-programming/property ) that issues a deprecation warning.

goodfeli commented on tests_tf/test_attacks.py:
```diff
@@ -1055,21 +1054,21 @@ def test_no_transformation(self):
-        new_labs = np.argmax(self.sess.run(self.model(x_adv)), axis=1)
-        self.assertTrue(np.mean(orig_labs == new_labs) < 0.7)
-        print(np.mean(orig_labs == new_labs) )
+    # def test_attack_strength(self):
```
Why is this test commented out? That looks scary

npapernot commented:
1. Since self.back is not used, I removed it.
2. I commented this test out because it is not implemented correctly (e.g., it passes a numpy array to a CleverHans model and then runs it in a session). I created #704 to make sure we don't forget to fix the test

goodfeli commented:
Thanks for removing self.back.

Regarding the test, two further thoughts:
1. I'd suggest just deleting it entirely rather than commenting it out. You could put a link to the commit that removes it in the issue, so that when someone works on the issue they can find the broken implementation quickly. My concern is that if we have a test for attack strength and we comment it out, it looks like the attack got weaker and we thought that was OK. I understand that that's not what happened, but we don't want to even look like we're doing that.
2. If this test wasn't implemented correctly, why wasn't it failing before? It sounds like it should have had some syntax error or raised an exception from passing the wrong type, etc.

npapernot commented:
Thanks for the feedback. I removed the test and added a pointer to the commit in the issue. I'm not sure why the test was not failing before (it was sometimes giving me a different outcome after running it multiple times).

goodfeli commented:
LGTM. Feel free to merge once the tests pass
